pkg/upgrade/upgrade.go

```diff
@@ -40,6 +40,35 @@ type JobDeps interface {
 	SystemDeps() SystemDeps
 }
 
+// RestoreBehavior explains what the behavior of a given upgrade migration is
+// during restoration of table, database or a cluster backups.
+//
+// For example, a migration that modifies every descriptor in the catalog e.g.
+// to populate a new field, needs to also insert itself into RESTORE, so that
+// when raw descriptors from a backup that could predate that migration, and
+// thus not have that field populated, are injected into the catalog, they are
+// migrated accordingly to uphold the invariant that all descriptors in the
+// catalog are migrated. Such a migration would use `RestoreActionImplemented`
+// and then a sentence specifying what is done during restore, for example:
+// RestoreActionImplemented("new field is populated during allocateDescriptorRequests").
+//
+// If an upgrade migration only affects state that is not restored, such as the
+// system tables that contain instance information and liveness, it can use a
+// RestoreActionNotRequired string explaining this fact, e.g.
+// `RestoreActionNotRequired("liveness table is not restored").
+type RestoreBehavior string
+
+// RestoreActionNotRequired is used to signal an upgrade migration only affects
+// persisted state that is not written by restore. The string should mention
+// why no action is required.
+type RestoreActionNotRequired = RestoreBehavior
+
+// RestoreActionImplemented is used to signal an upgrade migration affects state
+// that is written by restore and has implemented the requisite logic in restore
+// to perform the equivalent upgrade on restored data as it is restored. The
+// string should mention what is done during restore and where it is done.
```

Comment:
Re: B), I guess my concern was around someone updating a migration after it was originally written but before it's shipped in a release. I believe it's acceptable to modify a migration in that time frame right?

Re: C), ah makes sense, thanks for the context.

Reply:
Yeah, _currently_ we allow editing migrations while they live on master, before they go into a public release, but I suspect we might change that, since it is a dangerous habit to have, and we might be better served with a hard lint rule against any diffs that alter lines in pkg/upgrade/upgrades vs adding new files.

Anyway, I'm inclined to say that handling modifying is out of scope for now -- hopefully since modifications are only allowed immediately after it is written, prior to being released, then the fact we reminded the author when they wrote it to write the restore logic too means the fact the restore logic is there and needs to be updated as well should still be top of mind. I still think this change as-is is better than not having it.